File: backend/src/models/modelManager.js
const fs = require('fs');
const path = require('path');

/**
 * Discovers and manges model implementations.
 */
class ModelManager {
  constructor() {
    this.modelInstances = {};
    this.modelClasses = {};
    this.availableModels = [];
    this.discoveredModels = [];
    
    // Discover and load all model implementations
    this.discoverModels();
    
    // Apply environment variable filtering after discovering models
    this.filterModelsByEnvironment();
    
    console.log(`ModelManager initialized with available models: ${this.availableModels.join(', ')}`);
  }

  /**
   * Discover all models in the implementations directory
   */
  discoverModels() {
    const implementationsDir = path.join(__dirname, 'implementations');
    
    // Create the directory if it doesn't exist.
    // This is not a possibility since we already have sample implementations but just in case.
    if (!fs.existsSync(implementationsDir)) {
      fs.mkdirSync(implementationsDir);
      console.log('Created implementations directory');
      return;
    }
    
    const entries = fs.readdirSync(implementationsDir, { withFileTypes: true });
    
    const discoveredModels = [];
    
    for (const entry of entries) {
      const entryPath = path.join(implementationsDir, entry.name);
      
      try {
        let ModelClass;
        
        if (entry.isDirectory()) {
          // For directories, look for index.js as the entry point
          const indexPath = path.join(entryPath, 'index.js');
          if (fs.existsSync(indexPath)) {
            ModelClass = require(indexPath);
          } else {
            console.warn(`Skipping directory ${entry.name}: no index.js found`);
            continue;
          }
        } else if (entry.isFile() && entry.name.endsWith('.js')) {
          // For files, just require them directly
          ModelClass = require(entryPath);
        } else {
          // Skip non-js files
          continue;
        }
        
        // Create a temporary instance to get the model name
        if (typeof ModelClass === 'function') {
          const tempInstance = new ModelClass();
          if (typeof tempInstance.getModelName === 'function' || typeof tempInstance.getModelType === 'function') {
            const modelName = tempInstance.getModelName().toLowerCase();
            const modelType = tempInstance.getModelType();
            
            // Store the model class
            this.modelClasses[modelName] = ModelClass;
            discoveredModels.push({
              type: modelType,
              name: modelName,
            });
            
            console.log(`Discovered model: ${modelName} from ${entry.name}`);
          } else {
            console.warn(`Skipping ${entry.name}: missing getModelName() or getModelType() method.`);
          }
        } else {
          console.warn(`Skipping ${entry.name}: not a valid model class`);
        }
      } catch (error) {
        console.error(`Error loading model from ${entry.name}:`, error);
      }
    }
    
    this.discoveredModels = discoveredModels;
  }
  
  /**
   * Filter available models based on AVAILABLE_MODELS environment variable
   */
  filterModelsByEnvironment() {
    const whitelist = process.env.AVAILABLE_MODELS;
    
    if (!whitelist) {
      // No environment restriction, use all discovered models
      this.availableModels = [...this.discoveredModels];
      console.log('No AVAILABLE_MODELS environment variable set, using all discovered models');
      return;
    }
    
    const allowedModels = whitelist.split(',')
      .map(model => model.trim().toLowerCase())
      .filter(Boolean);
    
    // Filter discovered models to only include those in the allowed list
    this.availableModels = this.discoveredModels.filter(model => 
      allowedModels.includes(model.name)
    );
    
    if (this.availableModels.length === 0) {
      console.warn(`No models matched between discovered models and AVAILABLE_MODELS environment variable`);
      console.warn(`Discovered: ${this.discoveredModels.map(model => model.name).join(', ')}`);
      console.warn(`Allowed: ${allowedModels.join(', ')}`);
    } else {
      console.log(`Filtered models based on AVAILABLE_MODELS environment variable`);
      console.log(`Available models: ${this.availableModels.map(model => model.name).join(', ')}`);
    }
  }

  /**
   * Get a model instance by name
   * @param {string} modelName - Name of the model to get
   * @returns {Object} - Object containing either the model instance or an error
   */
  getModelByName(modelName) {
    if (!modelName) {
      return {
        success: false,
        error: 'No model specified',
        model: null
      };
    }
    
    const normalizedName = modelName.toLowerCase();
    
    if (!this.availableModels.map(model => model.name).includes(normalizedName)) {
      return {
        success: false,
        error: `Invalid model: ${modelName}. Available models: ${this.availableModels.join(', ')}`,
        model: null
      };
    }
    
    if (!this.modelInstances[normalizedName]) {
      try {
        this.modelInstances[normalizedName] = this.createModelInstance(normalizedName);
        console.log(`Created new instance of model: ${normalizedName}`);
      } catch (error) {
        console.error(`Error creating model instance for ${normalizedName}:`, error);
        return {
          success: false,
          error: `Error initializing model: ${error.message}`,
          model: null
        };
      }
    }
    
    return {
      success: true,
      error: null,
      model: this.modelInstances[normalizedName]
    };
  }
  
  /**
   * Get the list of available models
   * @returns {Array} - Array of available model names
   */
  getAvailableModels() {
    return this.availableModels.map(model => model.name);
  }
  
  /**
   * Get the list of models for type
   * @param {string} type - Type of the models to return 
   * @returns {Array} - Array of available model names
   */
  getAvailableModelsByType(type) {
    return this.availableModels.filter(model => model.type === type).map(model => model.name);
  }

  /**
   * Create a new model instance
   * @param {string} modelName - Name of the model to create
   * @returns {BaseModel} - A new instance of the requested model
   */
  createModelInstance(modelName) {
    const ModelClass = this.modelClasses[modelName];
    
    if (!ModelClass) {
      throw new Error(`Model class not found for: ${modelName}`);
    }
    
    return new ModelClass();
  }
}

const modelManager = new ModelManager();
module.exports = modelManager;

File: backend-py/src/models/modelManager.py
import os
import importlib
import inspect
import pkgutil
import sys
import logging
from typing import Dict, List, Optional, Any, Tuple

from src.models.baseModel import BaseModel

class ModelManager:
    """
    Discovers and manages model implementations.
    """
    
    def __init__(self):
        """
        Initialize the ModelManager
        """
        self.model_instances: Dict[str, BaseModel] = {}
        self.model_classes: Dict[str, type] = {}
        self.available_models: List[Dict[str, str]] = []
        self.discovered_models: List[Dict[str, str]] = []
        
        # Discover and load all model implementations
        self.discover_models()
        
        # Apply environment variable filtering after discovering models
        self.filter_models_by_environment()
        
        logging.info(f"ModelManager initialized with available models: {', '.join([model['name'] for model in self.available_models])}")
    
    def discover_models(self):
        """
        Discover all models in the implementations directory
        """
        discovered_models = []
        
        # Import the implementations package
        try:
            from src.models import implementations
            # Get the path to the implementations package
            implementations_path = os.path.dirname(implementations.__file__)
            
            # Walk through the implementations directory
            for _, module_name, is_pkg in pkgutil.iter_modules([implementations_path]):
                if module_name == '__pycache__':
                    continue
                
                try:
                    # Import the module
                    module = importlib.import_module(f"src.models.implementations.{module_name}")
                    
                    # Look for classes in the module that inherit from BaseModel
                    for name, obj in inspect.getmembers(module):
                        if (inspect.isclass(obj) and 
                            issubclass(obj, BaseModel) and 
                            obj != BaseModel):
                            
                            try:
                                # Create a temporary instance to get model info
                                temp_instance = obj()
                                model_name = temp_instance.get_model_name().lower()
                                model_type = temp_instance.get_model_type().value
                                
                                # Store the model class
                                self.model_classes[model_name] = obj
                                discovered_models.append({
                                    "type": model_type,
                                    "name": model_name,
                                })
                                
                                logging.info(f"Discovered model: {model_name} from {module_name}")
                            except Exception as e:
                                logging.warning(f"Failed to initialize model class {name}: {str(e)}")
                except Exception as e:
                    logging.error(f"Error loading model from {module_name}: {str(e)}")
        except Exception as e:
            logging.error(f"Error discovering models: {str(e)}")
        
        self.discovered_models = discovered_models
    
    def filter_models_by_environment(self):
        """
        Filter available models based on AVAILABLE_MODELS environment variable
        """
        whitelist = os.environ.get("AVAILABLE_MODELS")
        
        if not whitelist:
            # No environment restriction, use all discovered models
            self.available_models = self.discovered_models.copy()
            logging.info("No AVAILABLE_MODELS environment variable set, using all discovered models")
            return
        
        allowed_models = [model.strip().lower() for model in whitelist.split(",") if model.strip()]
        
        # Filter discovered models to only include those in the allowed list
        self.available_models = [
            model for model in self.discovered_models
            if model["name"] in allowed_models
        ]
        
        if not self.available_models:
            logging.warning("No models matched between discovered models and AVAILABLE_MODELS environment variable")
            logging.warning(f"Discovered: {', '.join([model['name'] for model in self.discovered_models])}")
            logging.warning(f"Allowed: {', '.join(allowed_models)}")
        else:
            logging.info("Filtered models based on AVAILABLE_MODELS environment variable")
            logging.info(f"Available models: {', '.join([model['name'] for model in self.available_models])}")
    
    def get_model_by_name(self, model_name):
        """
        Get a model instance by name
        
        Args:
            model_name (str): Name of the model to get
            
        Returns:
            dict: Object containing either the model instance or an error
        """
        if not model_name:
            return {
                "success": False,
                "error": "No model specified",
                "model": None
            }
        
        normalized_name = model_name.lower()
        
        if normalized_name not in [model["name"] for model in self.available_models]:
            return {
                "success": False,
                "error": f"Invalid model: {model_name}. Available models: {', '.join([model['name'] for model in self.available_models])}",
                "model": None
            }
        
        if normalized_name not in self.model_instances:
            try:
                self.model_instances[normalized_name] = self.create_model_instance(normalized_name)
                logging.info(f"Created new instance of model: {normalized_name}")
            except Exception as e:
                logging.error(f"Error creating model instance for {normalized_name}: {str(e)}")
                return {
                    "success": False,
                    "error": f"Error initializing model: {str(e)}",
                    "model": None
                }
        
        return {
            "success": True,
            "error": None,
            "model": self.model_instances[normalized_name]
        }
    
    def get_available_models(self):
        """
        Get the list of available models
        
        Returns:
            list: Array of available model names
        """
        return [model["name"] for model in self.available_models]
    
    def get_available_models_by_type(self, model_type):
        """
        Get the list of models for a specific type
        
        Args:
            model_type (str): Type of the models to return
            
        Returns:
            list: Array of available model names of the specified type
        """
        return [
            model["name"] for model in self.available_models 
            if model["type"] == model_type
        ]
    
    def create_model_instance(self, model_name):
        """
        Create a new model instance
        
        Args:
            model_name (str): Name of the model to create
            
        Returns:
            BaseModel: A new instance of the requested model
        """
        model_class = self.model_classes.get(model_name)
        
        if not model_class:
            raise ValueError(f"Model class not found for: {model_name}")
        
        return model_class()

# Create a singleton instance
model_manager = ModelManager()
